File: new-resource-detector/src/main/java/nl/haploid/resource/detector/service/ResourceProducerService.java
package nl.haploid.resource.detector.service;

import nl.haploid.event.JsonMapper;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.clients.producer.RecordMetadata;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;

@Service
public class ResourceProducerService {

    @Autowired
    private KafkaProducer<String, String> kafkaProducer;

    @Autowired
    private JsonMapper jsonMapper;

    // TODO: unit test/IT
    public Future<RecordMetadata> publishResourceDescriptor(final ResourceDescriptor descriptor) {
        final String topic = "TODO"; // TODO: topic name
        final String message = jsonMapper.toString(descriptor);
        return kafkaProducer.send(new ProducerRecord<>(topic, message));
    }

    // TODO: unit test/IT
    public RecordMetadata resolveFuture(final Future<RecordMetadata> future) {
        try {
            return future.get();
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException("Could not resolve future; message may not have been produced!", e);
        }
    }
}

File: resource-builder/src/main/scala/nl/haploid/octowight/builder/service/ResourceProducerService.scala
package nl.haploid.octowight.builder.service

import java.util.concurrent.{Future, TimeUnit}

import nl.haploid.octowight.JsonMapper
import nl.haploid.octowight.kafka.producer.KafkaOutChannel
import nl.haploid.octowight.registry.data.ResourceMessage
import org.apache.kafka.clients.producer.{ProducerRecord, RecordMetadata}
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.{Autowired, Value}
import org.springframework.stereotype.Service

@Service
class ResourceProducerService extends KafkaOutChannel {
  private[this] lazy val log = LoggerFactory.getLogger(getClass)

  @Value("${octowight.kafka.topic.resources.built}") val topic: String = null
  @Autowired private[this] val jsonMapper: JsonMapper = null

  def send(resources: Iterable[ResourceMessage]): Unit = {
    val futures: Iterable[Future[RecordMetadata]] = resources map {
      resource => {
        val message = jsonMapper.serialize(resource)
        log.debug(s"Send message to $topic: $message")
        kafkaProducer.send(new ProducerRecord[String, String](topic, message))
      }
    }
    futures.map(_.get(5, TimeUnit.SECONDS))
  }
}
